src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java

```diff
@@ -14,30 +14,158 @@
+    datastore.put(lectureEntity);
+    response.sendRedirect(buildRedirectUrl(lectureEntity));
   }
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.sendError(500, "Not Implemented");
+    List<Lecture> lectures = getLectures();
+    Gson gson = new Gson();
+    response.setContentType("application/json");
+    response.getWriter().println(gson.toJson(lectures));
+  }
+
+  /**
+   * Returns the Entity in database that has {@code url}, or
+   * {@code Optional.empty()} if one doesn't exist.
+   */
+  private Optional<Entity> checkUrlInDatabase(String url) {
+    Query query = new Query(Lecture.ENTITY_KIND);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+    Iterable<Entity> resultsIterable = results.asIterable();
+
+    for (Entity lecture : resultsIterable) {
+      if (lecture.getProperty(Lecture.PROP_URL).equals(url)) {
+        return Optional.of(lecture);
+      }
+    }
+    return Optional.empty();
+  }
+
+  /** Creates and returns {@code lectureEntity} using parameters found in {@code request}. */
+  private Entity createLectureEntity(HttpServletRequest request) {
+    String lectureName = getParameter(request, NAME_INPUT, DEFAULT_VALUE);
+    String videoUrl = getParameter(request, LINK_INPUT, DEFAULT_VALUE);
+    String videoId = getVideoId(videoUrl);
+
+    Entity lectureEntity = new Entity(Lecture.ENTITY_KIND);
+    lectureEntity.setProperty(Lecture.PROP_NAME, lectureName);
+    lectureEntity.setProperty(Lecture.PROP_URL, videoUrl);
+    lectureEntity.setProperty(Lecture.PROP_VIDEO_ID, videoId);
+    return lectureEntity;
+  }
+
+  /** Returns a list of lectures stored in the database. */
+  private List<Lecture> getLectures() {
+    Query query = new Query(Lecture.ENTITY_KIND);
+    PreparedQuery results = datastore.prepare(query);
+    List<Lecture> lectures = new ArrayList<>();
+    for (Entity lecture : results.asIterable()) {
+      Lecture newLecture = Lecture.fromLectureEntity(lecture);
+      lectures.add(newLecture);
+    }
+    return lectures;
+  }
+
+  /**
+   * Returns value with {@code name} from the {@code request} form.
+   * If the {@code name} cannot be found, return {@code defaultValue}.
+   */
+  private String getParameter(HttpServletRequest request, String name, String defaultValue) {
+    String value = request.getParameter(name);
+    if (value == null) {
+      return defaultValue;
+    }
+    return value;
+  }
+
+  /** Returns YouTube video ID for a given {@code videoUrl}. */
+  private String getVideoId(String videoUrl) {
+    Matcher matcher = videoUrlGeneratedPattern.matcher(videoUrl);
+    if (matcher.find()) {
+      return matcher.group();
+    }
+    // TODO: Throw an error saying ID not found.
+    return "";
+  }
+
+  /**
+   * Returns URL to redirect to with parameters {@code lectureId} and {@code videoId}
```

Comment:
Returns URL to redirect to... 

Redirect to what?

Reply:
fixed